presto-main/src/main/java/com/facebook/presto/operator/SetBuilderOperator.java

```diff
@@ -203,12 +225,39 @@ public void addInput(Page page)
 
         Block sourceBlock = page.getBlock(setChannel);
         Page sourcePage = hashChannel.isPresent() ? new Page(sourceBlock, page.getBlock(hashChannel.get())) : new Page(sourceBlock);
-        channelSetBuilder.addPage(sourcePage);
+        unfinishedWork = channelSetBuilder.addPage(sourcePage);
+        processUnfinishedWork();
+        channelSetBuilder.updateMemoryReservation();
```

Comment:
How many times per page `OperatorContext#setMemoryReservation` is being called now?

Reply:
twice, one in needsInput(), one in addInput()

Reply:
Actually it's also called in BigintGroupByHash.java.tryRehash() as a callback function. But just calling there is not enough since there are other memory growth on the hash table happening in between, so we need to explicitly call it in SetBuildOperator class.